citest/gcp_testing/google_cloud_storage_agent.py

```diff
@@ -0,0 +1,126 @@
+from ..service_testing import base_agent
+
+
+READ_ONLY_SCOPE = 'https://www.googleapis.com/auth/devstorage.read_only'
+READ_WRITE_SCOPE = 'https://www.googleapis.com/auth/devstorage.read_write'
+FULL_SCOPE = 'https://www.googleapis.com/auth/devstorage.full_control'
+
+
+class GoogleCloudStorageAgent(base_agent.BaseAgent):
+  """Agent that interacts with Google Cloud Storage service."""
+
+  def __init__(self, json_credential_path, scope):
+    """Constructs agent.
+
+    Args:
+      json_credential_path: [String] path to JSON credentials for
+          service account.
+    """
+    super(GoogleCloudStorageAgent, self).__init__()
+    self.__service = build_authenticated_service(
+        'storage', 'v1', scope, json_credential_path)
+
+  def inspect(self, bucket, path=None, generation=None):
+    """Get metadata for a bucket or object in the bucket.
+
+    Args:
+      bucket: [string] The bucket to inspect.
+      path: [string] The name of the object in the bucket.
+         If None then inspect the bucket itself.
+
+    Returns:
+      Metadata for specified resource.
+    """
+    if not path:
+      self.logger.info('Inspecting bucket=%s', bucket)
+      request = self.__service.buckets().get(bucket=bucket)
+    else:
+      maybe_gen = ' generation={0}'.format(generation) if generation else ''
+      self.logger.info('Inspecting bucket=%s, path=%s%s',
+                       bucket, path, maybe_gen)
+      request = self.__service.objects().get(bucket=bucket, object=path,
+                                             generation=generation)
+    response = request.execute()
+    self.logger.info('%s', response)
+    return response
+
+  def list(self, bucket, path, with_versions, fields=None):
+    """List the contents of the path in the specified bucket.
+
+    Args:
+      bucket: [string] The name of the bucket to list.
+      path: [string] The path within the bucket to list.
+         If path is a directory, it will list the immediate contents, but
+         not content of any subdirectories.
+      with_versions: [boolean] Whether or not to list all the versions.
+         If path is a directory, this will list all the versions of all the
+         files. Otherwise just all the versions of the given file.
+      fields: [string] A comma-delimited selector indicating the subset of
+         fields to include as partial responses, or None for everything.
+
+    Returns:
+      A list of resources.
+    """
+    self.logger.info('Listing bucket=%s prefix=%s versions=%s',
+                     bucket, path, with_versions)
+    request = self.__service.objects().list(bucket=bucket,
+                                            prefix=path, fields=fields,
+                                            versions=with_versions)
+    all_objects = []
+    while request:
+      response = request.execute()
+      all_objects.extend(response.get('items', []))
+      request = self.__service.objects().list_next(request, response)
```

Comment:
Only by increasing the result limit. This is chaining together pages, which is inherently serial.